Autotools compliance: out of source build fails on Windows 10 mingw64

## Environment

Windows 10 mingw64 - newly updated. Source from git 

```sh
zzzap@LISTER MINGW64 ~/local-code-projects/my-code/roswell
$ git log -n 1
commit 05a4556d6dc01bda227a4325cc80dd89b7f863e4 (HEAD -> master, tag: v19.3.10.97, upstream/release, upstream/master, origin/master, origin/HEAD)
Author: SANO Masatoshi <snmsts@gmail.com>
Date:   Tue Mar 5 17:25:34 2019 +0800

    bump version to 19.3.10.97

zzzap@LISTER MINGW64 ~/local-code-projects/my-code/roswell
$ uname -a
MINGW64_NT-10.0 LISTER 2.11.2(0.329/5/3) 2018-11-26 09:22 x86_64 Msys

zzzap@LISTER MINGW64 ~/local-code-projects/my-code/roswell
```


## Expected behaviour

in source builds, completes compile, runs ros --version and run config.

out of source build, should compile same as in source build.

## Actual result

In source build correctly compiles, installs ros, ros --version and ros config run
```sh
zzzap@LISTER MINGW64 ~/local-code-projects/my-code/roswell
$ ros --version
roswell 19.3.10.97(05a4556)

zzzap@LISTER MINGW64 ~/local-code-projects/my-code/roswell
$ ros config
Installing sbcl-bin...
[##########################################################################]100%
[##########################################################################]100%Install Script for sbcl-bin...
Installing Quicklisp...
up to date. stop
No SBCL version specified. Downloading platform-table.html to see the available versions...
Installing sbcl-bin/1.4.14...
Downloading https://github.com/roswell/sbcl_bin/releases/download/1.4.14/sbcl-1.4.14-x86-64-windows-binary.msi
Extracting the msi archive. sbcl-bin-1.4.14-x86-64-windows.msi to C:\Users\zzzap\.roswell\src\sbcl-1.4.14-x86-64-windows\

Making core for Roswell...
up to date. stop
setup.time=3761596180
sbcl-bin.version=1.4.14
default.lisp=sbcl-bin

Possible subcommands:
set
show

zzzap@LISTER MINGW64 ~/local-code-projects/my-code/roswell
$
```

Out of source build has a error.

```sh
gcc -DHAVE_CONFIG_H -I. -I../../src -I..     -g -O2 -MT cmd-internal.o -MD -MP -MF .deps/cmd-internal.Tpo -c -o cmd-internal.o cmd-internal.c
gcc.exe: error: cmd-internal.c: No such file or directory
gcc.exe: fatal error: no input files
compilation terminated.
make[2]: *** [Makefile:502: cmd-internal.o] Error 1
make[2]: Leaving directory '/c/Users/zzzap/local-code-projects/my-code/roswell/build/src'
make[1]: *** [Makefile:555: all-recursive] Error 1
make[1]: Leaving directory '/c/Users/zzzap/local-code-projects/my-code/roswell/build'
make: *** [Makefile:413: all] Error 2
```

## What terms "in source" and "out of source" mean

in source - make is executed within the source tree directory.
out of source - is executed in a new empty directory

out of source build is expected from an autotools system.

Following scripts show build command for each:

Roswell build in source
-----------------------
Build tree within source tree.

```sh
sh bootstrap
./configure
make
make install
```

This is a sucess, ros -version and ros config both work.

Roswell build out of source
---------------------------
Build tree separate from source tree.
```sh
sh bootstrap
mkdir build
cd build
..\configure
make
make install
```

## Summary

Out of source build is expected with autotools. Roswell does build in source but fails out of source build.

--- a/Makefile.am
+++ b/Makefile.am
@@ -1,5 +1,5 @@
-roslispdir = $(sysconfdir)/$(PACKAGE)
-roslisppatchdir = $(sysconfdir)/$(PACKAGE)/patch
+roslispdir = @roslispdir@
+roslisppatchdir = @roslisppatchdir@
 SUBDIRS = src
 
 if MANUAL_INSTALL
@@ -92,19 +92,6 @@ roslisp_DATA = $(LISPS:%=lisp/%)
 
 roslisppatch_DATA = patch/sbcl-1.3.11.patch patch/sbcl-posix-tests.patch
 
-src/gend.h: FORCE
-	printf '#define ROS_COMPILE_ENVIRONMENT "%s"\n' "`$(CC) --version|head -n 1`" > $@.tmp
-	(printf "#define ROS_REVISION \"" && ((which git>/dev/null&&[ -e .git ]&& \
-	(git log -n 1 --oneline|cut -d' ' -f1| tr -d '\n'| tr -d '\r'))||printf "") && printf "\"\n") >> $@.tmp
-	printf '#define PATCH_PATH "%s"\n' "$(roslisppatchdir)" >> $@.tmp
-if WITH_WINDOWS
-	printf '#define LISP_PATH "%s"\n' "`mkdir -p $(roslispdir);cd $(roslispdir); pwd -W 2>/dev/null`" >> $@.tmp
-else
-	printf '#define LISP_PATH "%s"\n' "$(roslispdir)" >> $@.tmp
-endif
-	cmp -s $@.tmp $@||cp $@.tmp $@
-	cat $@
-	rm -f $@.tmp
 
 release-prepare:
 	ros scripts/release.ros prepare

--- a/configure.ac
+++ b/configure.ac
@@ -35,10 +35,17 @@ AM_CONDITIONAL([WITH_WINDOWS], [test "$with_windows" != ""])
 AC_CHECK_TOOL([WINDRES], [windres], [])
 AM_CONDITIONAL([WITH_WIN_ICON], [test "$WINDRES" != ""])
 
-# This code needs git
-# Check for git source version at config time
-AC_SUBST([git_version], m4_esyscmd_s([git describe |  sed -n -e 's/^.*-g//p'],[git SHA1 for HEAD.])) 
-AC_SUBST([git_version_long], m4_esyscmd_s([git describe --abbrev=7 --dirty]), [long version for git source])
+# Git is optional, so fail gracefully, build from tar may not have .git/
+AC_CHECK_PROG(GIT_CHECK,git,yes)
+AS_IF([test "x$GIT_CHECK" = xyes],
+      [AC_SUBST([git_version], `set -o pipefail ; { git describe  | sed -n -e 's/^.*-g//p' ; } 2>/dev/null || echo 'NO-GIT-REVISION'
+`,[git SHA1 for HEAD.])],
+      [AC_SUBST([git_version],[NO-GIT-REVISION])])
+      
+AS_IF([test "x$GIT_CHECK" = xyes],      
+      [AC_SUBST([git_version_long], `set -e | git describe --abbrev=7 --dirty 2>/dev/null || echo 'NO-GIT-REVISION'`, [long version for git source])],
+      [AC_SUBST([git_version_long], [NO-GIT-REVISION])])
+
 
 # Define rosdir and rospatchdir
 # TODO: Make these ./configure line options.

--- a/configure.ac
+++ b/configure.ac
@@ -4,6 +4,8 @@
 AC_PREREQ([2.59])
 AC_INIT([roswell],[19.3.10.97],snmsts@gmail.com)
 
+
+
 AM_CONFIG_HEADER(config.h)
 AC_CONFIG_FILES([Makefile src/Makefile])
 AC_CONFIG_SRCDIR([Makefile.am])
@@ -32,6 +34,20 @@ AC_CHECK_HEADERS(windows.h,[with_windows=true])
 AM_CONDITIONAL([WITH_WINDOWS], [test "$with_windows" != ""])
 AC_CHECK_TOOL([WINDRES], [windres], [])
 AM_CONDITIONAL([WITH_WIN_ICON], [test "$WINDRES" != ""])
+
+# This code needs git
+# Check for git source version at config time
+AC_SUBST([git_version], m4_esyscmd_s([git describe |  sed -n -e 's/^.*-g//p'],[git SHA1 for HEAD.])) 
+AC_SUBST([git_version_long], m4_esyscmd_s([git describe --abbrev=7 --dirty]), [long version for git source])
+
+# Define rosdir and rospatchdir
+# TODO: Make these ./configure line options.
+# See https://www.gnu.org/prep/standards/html_node/Directory-Variables.html#Directory-Variables
+# Especially 'lispdir' example for emacs.
+roslispdir='${sysconfdir}/roswell'
+AC_SUBST(roslispdir)
+roslisppatchdir='${roslispdir}/patch'
+AC_SUBST(roslisppatchdir)
 
 # Change master uri for sbcl-bin
 AC_ARG_WITH([sbcl_bin_base],

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -7,8 +7,20 @@ html_sbcl_SOURCES =  html.c html-sbcl-bin.c main.c opt.c $(ros_utils)
 html_sbcl_CFLAGS = "-D ROSWELL_HTML_TEST"
 
 gend.h: FORCE
-	make -C .. src/gend.h
-cmd-internal.c: gend.h
+	printf '#define ROS_COMPILE_ENVIRONMENT "%s"\n' "`$(CC) --version|head -n 1`" > $@.tmp
+	printf '#define ROS_REVISION "@git_version@"\n' >> $@.tmp
+	printf '#define PATCH_PATH "@roslisppatchdir@"\n' >> $@.tmp
+if WITH_WINDOWS
+	printf '#define LISP_PATH "%s"\n' "`mkdir -p @roslispdir@;cd @roslispdir@; pwd -W 2>/dev/null`" >> $@.tmp
+else
+	printf '#define LISP_PATH "@roslispdir@"\n' >> $@.tmp
+endif
+	cmp -s $@.tmp $@||cp $@.tmp $@
+	cat $@
+	rm -f $@.tmp
+
+cmd-internal.o: gend.h
+
 bin_PROGRAMS = ros
 ros_SOURCES = main.c opt.c download.c download_windows.c archive.c archive_windows.c \
 	register-commands.c html.c html-sbcl-bin.c \

